packages/aws-cdk/test/api/state-machine-hotswap-deployments.test.ts

```diff
@@ -0,0 +1,226 @@
+import { MockSdkProvider } from '../util/mock-sdk';
+import { FakeCloudformationStack } from './fake-cloudformation-stack';
+
+const STACK_NAME = 'withouterrors';
+const STACK_ID = 'stackId';
+
+let mockSdkProvider: MockSdkProvider;
+let mockUpdateMachineCode: (params: StepFunctions.Types.UpdateStateMachineInput) => StepFunctions.Types.UpdateStateMachineOutput;
+let mockListStackResources: (params: CloudFormation.Types.ListStackResourcesInput) => CloudFormation.Types.ListStackResourcesOutput;
+let currentCfnStack: FakeCloudformationStack;
+
+beforeEach(() => {
+  jest.resetAllMocks();
+  mockSdkProvider = new MockSdkProvider({ realSdk: false });
+  mockListStackResources = jest.fn(() => {
+    let summaries: StackResourceSummaries = [];
+
+    // Mock state machine stack resource
+    summaries[0] = {
+      LogicalResourceId: 'Machine',
+      ResourceType: 'AWS::StepFunctions::StateMachine',
+      ResourceStatus: 'CREATE_COMPLETE',
+      LastUpdatedTimestamp: new Date(),
+      PhysicalResourceId: 'mock-machine-resource-id',
+    };
+
+    return { StackResourceSummaries: summaries };
+  });
+  mockUpdateMachineCode = jest.fn();
+  mockSdkProvider.stubCloudFormation({
+    listStackResources: mockListStackResources,
+  });
+  mockSdkProvider.stubStepFunctions({
+    updateStateMachine: mockUpdateMachineCode,
+  });
+  currentCfnStack = new FakeCloudformationStack({
+    stackName: STACK_NAME,
+    stackId: STACK_ID,
+  });
+});
+
+test('returns undefined when a new StateMachine is added to the Stack', async () => {
+  // GIVEN
+  const cdkStackArtifact = cdkStackArtifactOf({
+    template: {
+      Resources: {
+        Machine: {
+          Type: 'AWS::StepFunctions::StateMachine',
+        },
+      },
+    },
+  });
+
+  // WHEN
+  const deployStackResult = await tryHotswapDeployment(mockSdkProvider, {}, currentCfnStack, cdkStackArtifact);
+
+  // THEN
+  expect(deployStackResult).toBeUndefined();
+});
+
+test('calls the updateStateMachine() API when it receives only a definitionString change in a state machine', async () => {
+  // GIVEN
+  currentCfnStack.setTemplate({
+    Resources: {
+      Machine: {
+        Type: 'AWS::StepFunctions::StateMachine',
+        Properties: {
+          DefinitionString: {
+            'Fn::Join': [
+              '',
+              [
+                '{ "Prop" : "old-value", "AnotherProp" : "another-old-value" }',
+              ],
+            ],
+          },
+          StateMachineName: 'my-machine',
+        },
+      },
+    },
+  });
+  const cdkStackArtifact = cdkStackArtifactOf({
+    template: {
+      Resources: {
+        Machine: {
+          Type: 'AWS::StepFunctions::StateMachine',
+          Properties: {
+            DefinitionString: {
+              'Fn::Join': [
+                '',
+                [
+                  '{ "Prop" : "new-value", "AnotherProp" : "another-new-value" }',
+                ],
+              ],
+            },
```

Comment:
Yeah... I'm going to need to see a passing test where `DefinitionString` is just `'{ "Prop" : "new-value", "AnotherProp" : "another-new-value" }'`, or something similar, before I give my ShipIt 😜.